### PR TITLE
Turn incorrect use of input annotation into an error

### PR DESCRIPTION
**Pull request by melix**
This commit also introduces a dummy annotation, `@ValidationProblem`, which
can be used in tests to trigger a validation problem. This was required because
some tests actually verify the behavior of Gradle when a validation problem
happens but that the problem is not fatal (execution continues). If we turn
all validation problems into errors, then we don't have any annotation to use
in tests.

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/BuildResultLoggerIntegrationTest.groovy [hunk 1]
```diff
@@ -94,10 +96,14 @@ class BuildResultLoggerIntegrationTest extends AbstractIntegrationSpec implement
         !output.contains("actionable tasks")
     }
 
+    @IgnoreIf({ !GradleContextualExecuter.embedded })
```
Comment (wolfs) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/BuildResultLoggerIntegrationTest.groovy:99]:
```suggestion
    @Requires({ GradleContextualExecuter.embedded })
```

Note that you need to use Spock's `@Require`s, not Gradle's.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputFilePropertiesIntegrationTest.groovy [hunk 2]
```diff
@@ -157,29 +157,6 @@ class TaskInputFilePropertiesIntegrationTest extends AbstractIntegrationSpec imp
         executed ":foo"
     }
 
-    @Issue("https://github.com/gradle/gradle/issues/9674")
-    def "allows @Input of task with no actions to be null"() {
-        buildFile << """
-            class FooTask extends DefaultTask {
-               @Input
```
Comment (wolfs) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputFilePropertiesIntegrationTest.groovy:164]: Let's replace `@Input` with `@InputFiles` instead. That should expose the problem of https://github.com/gradle/gradle/issues/9674 without running into a validation problem.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/InputPropertyAnnotationHandler.java [hunk 3]
```diff
@@ -65,9 +67,15 @@ public void validatePropertyMetadata(PropertyMetadata propertyMetadata, TypeVali
...
             || FileCollection.class.isAssignableFrom(valueType)) {
-            validationContext.visitPropertyProblem(WARNING,
-                propertyMetadata.getPropertyName(),
-                String.format("has @Input annotation used on property of type '%s'", ModelType.of(valueType).getDisplayName())
+            validationContext.visitPropertyProblem(problem ->
+                problem.withId(ValidationProblemId.INCORRECT_USE_OF_INPUT_ANNOTATION)
+                .forProperty(propertyMetadata.getPropertyName())
+                .reportAs(ERROR)
+                .withDescription(() -> String.format("has @Input annotation used on property of type '%s'", ModelType.of(valueType).getDisplayName()))
+                .happensBecause(() -> "A property of type '" + ModelType.of(valueType).getDisplayName() + "' is annotated with @Input cannot determine how to interpret the file")
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/InputPropertyAnnotationHandler.java:75]:
```suggestion
                .happensBecause(() -> "A property of type '" + ModelType.of(valueType).getDisplayName() + "' 
 annotated with @Input cannot determine how to interpret the file")
```

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/InputPropertyAnnotationHandler.java [hunk 4]
```diff
@@ -65,9 +67,15 @@ public void validatePropertyMetadata(PropertyMetadata propertyMetadata, TypeVali
...
+                .addPossibleSolution("Annotate with @InputFile for regular files")
+                .addPossibleSolution("Annotate with @InputDirectory for directories")
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/InputPropertyAnnotationHandler.java:77]: If you want to track the path, you can also return `File.absolutePath`, return a `String` and keep the `@Input`.

File: subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc [hunk 5]
```diff
@@ -200,4 +200,13 @@ Because there are no semantics associated with this combination, Gradle cannot d
...
 To fix this, you most likely need to remove the conflicting modifier annotation or check that the actual property type is what you intended.
 
+[[incorrect_use_of_input_annotation]]
+== Incorrect use of the `@Input` annotation
+
+This error indicates that a property is annotated with `@Input`, but that it should be annotated with `@InputFile` or `@InputDirectory` instead.
+
+If you use the `@Input` annotation on a file-based property, Gradle wouldn't consider the file contents, or the directory contents, as inputs, as you might expect.
+
+To fix this problem, you need to tell Gradle if the file property represents an input file, in which case you should annotate it with `@InputFile`, or a directory, in which case it should be annotated with `@InputDirectory`.
```
Comment (wolfs) [inline subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc:210]: We should fix the Javadoc on `@Input`, `@InputFile` and `@InputDirectory` as well. See https://github.com/gradle/gradle/issues/5789

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 6]
```diff
@@ -101,6 +102,7 @@
         .displayName("Global services")
         .parent(newCommandLineProcessLogging())
         .parent(NativeServicesTestFixture.getInstance())
+        .parent(ValidationServicesFixture.getServices())
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:105]: Great that this works! Does this have any side-effect for the non-embedded tests?

Comment (melix) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:105]: It doesn't, as far as I can tell. I wish it worked for non-embedded tests too, that would have been even better!

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/internal/reflect/validation/ValidationMessageChecker.groovy [hunk 7]
```diff
@@ -168,15 +212,32 @@ trait ValidationMessageChecker {
...
             super(checker)
+            forWorkItem()
         }
 
         AnnotationContext annotation(String name) {
             annotation = name
             this
         }
+
+        AnnotationContext forInjection() {
```
Comment (wolfs) [inline subprojects/model-core/src/testFixtures/groovy/org/gradle/internal/reflect/validation/ValidationMessageChecker.groovy:227]: This is for the transform action.

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/internal/reflect/validation/ValidationMessageChecker.groovy [hunk 8]
```diff
@@ -168,15 +212,32 @@ trait ValidationMessageChecker {
...
+            validAnnotations = "@Inject, @InputArtifact or @InputArtifactDependencies"
+            this
+        }
+
+        AnnotationContext forWorkItem() {
```
Comment (wolfs) [inline subprojects/model-core/src/testFixtures/groovy/org/gradle/internal/reflect/validation/ValidationMessageChecker.groovy:232]: This is for the transform parameters

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/internal/reflect/validation/ValidationMessageChecker.groovy (hunk 7 above)
Comment (melix) [reply subprojects/model-core/src/testFixtures/groovy/org/gradle/internal/reflect/validation/ValidationMessageChecker.groovy:227]: Will fix in upcoming PR.

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/internal/reflect/validation/ValidationMessageChecker.groovy (hunk 8 above)
Comment (melix) [reply subprojects/model-core/src/testFixtures/groovy/org/gradle/internal/reflect/validation/ValidationMessageChecker.groovy:232]: Will fix in upcoming PR.

**Pull request merged**